### ISSUE TITLE
Crash in native code while asynchronously loading TextureAtlas

### ISSUE THREAD

**Issue by justinbowes** [libgdx/libgdx#2763]
I'm seeing a crash when loading a TextureAtlas through the AssetManager that does not occur if I load from the main thread. (OSX 10.10, other platforms not checked yet.)

The TextureAtlas is multiple pages of PNGs, 1024x1024, generated by TexturePacker. The crash occurs on the second page it tries to load (page 8 of 9).

I can add an error log, but it's huge, so here are the parts I found interesting:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0xa) at pc=0x00007fff8973e512, pid=65013, tid=122375
#
# JRE version: Java(TM) SE Runtime Environment (8.0_25-b17) (build 1.8.0_25-b17)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.25-b02 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libsystem_platform.dylib+0x1512]  _platform_memmove$VARIANT$Nehalem+0x1d2
```

```
Stack: [0x000000012701e000,0x000000012711e000],  sp=0x000000012710ff40,  free space=967k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [libsystem_platform.dylib+0x1512]  _platform_memmove$VARIANT$Nehalem+0x1d2
C  [libgdx64.dylib+0x24b19]  stbi__getn+0x99
C  [libgdx64.dylib+0x23e64]  stbi__parse_png_file+0x974
C  [libgdx64.dylib+0x6b83]  stbi_load_main+0x203
C  [libgdx64.dylib+0x14dd2]  gdx2d_load+0x72
C  [libgdx64.dylib+0x14ce]  Java_com_badlogic_gdx_graphics_g2d_Gdx2DPixmap_load+0x3e
j  com.badlogic.gdx.graphics.g2d.Gdx2DPixmap.load([J[BII)Ljava/nio/ByteBuffer;+0
j  com.badlogic.gdx.graphics.g2d.Gdx2DPixmap.<init>([BIII)V+19
j  com.badlogic.gdx.graphics.Pixmap.<init>(Lcom/badlogic/gdx/files/FileHandle;)V+24
```

```
Registers:
RAX=0x000000012341c000, RBX=0x00000000000b8000, RCX=0x000000000000000f, RDX=0xffffffffffe46fb0
RSP=0x000000012710ff40, RBP=0x000000012710ff40, RSI=0x0000000780b09c4d, RDI=0x0000000123263030
R8 =0x0000000000000153, R9 =0x00000000000000cc, R10=0x0000000000000000, R11=0x000000012341bfff
R12=0x00000000ffffffff, R13=0x000000012711d208, R14=0x000000012341c000, R15=0xffffffffffffffff
RIP=0x00007fff8973e512, EFLAGS=0x0000000000010282, ERR=0x0000000000000006
  TRAPNO=0x000000000000000e
```

I understand this as [stb_image trying to parse an IDAT chunk](https://github.com/nothings/stb/blob/ec2158386c877a2388038d0587f22b129a2cc03e/stb_image.h#L4282). My calling-convention-fu is weak but if RDX is supposed to be parameter 3 to stbi__getn, that is suspect. 

I considered a corrupt PNG (generated by TexturePacker), or a failed realloc, but the latter should have already been handled by stb_image, and if I load the TextureAtlas on the main thread, there is no error.

Any ideas?

**Comment by justinbowes** [comment]
The TexturePacker task is running during the bootstrap of DesktopLauncher. If I disable the TexturePacker task, the crash never occurs. If I have it write fewer atlas pages, the crash rarely occurs. If this is a race related to file content flushing, I'm not sure it's worth a further look.

**Comment by justinbowes** [comment]
Happens with decreasing frequency as I decrease the number of texture pages. Closed because I am using a workaround (loading TextureAtlas in main thread) and I'm unlikely to submit a PR.

**Comment by badlogic** [comment]
Sounds like a concurrency issue to me. Don't start the LwjglApp before
TexturePacker is done.

**Comment by justinbowes** [comment]
Here's a gist I created to show you why I'm pulling my hair out.
[Minimal test app gist](https://gist.github.com/justinbowes/9ce01c14fc33c6bd258f)

Here is what I observe:
 # TexturePacker runs, then I `System.exit(0)`.
 # I load all the images that it generates. Synchronously, i.e. `new Texture(path)`. Usually there will be a single atlas page that causes a crash when loaded. If so, it is consistent -- it will always cause a crash until I next run TexturePacker. I can comment out all the other loads, and loading only the questionable file will crash.
 # Worse: The file looks fine. I can look at it in any image editing program. Only libgdx seems to have a problem reading it.
 # Even worse: I can copy the file and load the copy. No problems. Then I do a binary comparison, and they're identical. Or I can rename the file and it loads fine. Or I can rename the file, rename it back to its original filename, load it again...loads fine. 

In other words, it looks corrupt, but only to libgdx, and if I do so much as rename it, the problem disappears. I'm suspecting some kind of interaction between the JVM and the operating system, or maybe a hardware issue. 

I'll try it on a different machine, but honestly I was just hoping to whip up a prototype today.

**Comment by justinbowes** [comment]
Not reproducible on a different machine, same OS (10.10.1) and JDK (8u31).

**Comment by NathanSweet** [comment]
It's almost certainly not a libgdx problem if simply renaming the file "fixes" it. Maybe your IDE is copying the image into a bin directory and loading it thru the classpath from there. Maybe that copy is done while the image is being written.

**Comment by justinbowes** [comment]
Great thought. Where a PNG causes a crash in stb_image, the file included in the build is different than the one in android/assets. This reminded me of a difference in Eclipse configuration between the machine that failed and the one that succeeded.

The culprit is the Eclipse setting, `General > Workspace > Refresh using native hooks or polling`. When checked, the file is copied while being written. When unchecked (and Eclipse is restarted), the error does not occur.

Thanks @badlogic and @NathanSweet. Today is going to be a better day. :+1:

**Comment by NathanSweet** [comment]
For my projects I use `-Declipse` when starting the JVM, then when loading assets if `System.getProperty("eclipse") != null` (usually stored some static `eclipse` boolean field) I use an `absolute` FileHandle to load the file directly, otherwise I use `internal` to load thru the classpath. This way I never have IDE refresh shenanigans.

**Comment by justinbowes** [comment]
Implemented as you suggested in a `FileHandleResolver`. Thanks again!
